components/salsa-2022-macros/src/input.rs

```diff
@@ -198,11 +217,86 @@ impl InputStruct {
                     #(#field_getters)*
 
                     #(#field_setters)*
+
+                    #builder_fn
                 }
             }
         }
+    }
+
+    /// generate builder struct
+    fn input_builder(&self) -> (syn::ItemStruct, syn::ItemImpl) {
+        let ident = self.id_ident();
+        let struct_name = self.builder_name();
+        let field_names = self.all_field_names();
+        let field_tys: Vec<_> = self.all_field_tys();
+        let db_dyn_ty = self.db_dyn_ty();
+        let jar_ty = self.jar_ty();
+        let input_index = self.input_index();
+        let field_indices = self.all_field_indices();
+
+        let buidler: syn::ItemStruct = parse_quote! {
+            struct #struct_name{
+                #(#field_names: Option<#field_tys>,)*
+                durability: Option<salsa::durability::Durability>,
+            }
+        };
+
+        let build_fn: syn::ImplItemMethod = if self.0.is_isingleton() {
+            parse_quote! {
+                fn build(&mut self, __db: &#db_dyn_ty) -> std::result::Result<#ident, std::boxed::Box<dyn std::error::Error>> {
+                    let durability = self.durability.ok_or_else(|| "durability not provided")?;
 
-        // }
+                    let (__jar, __runtime) = <_ as salsa::storage::HasJar<#jar_ty>>::jar(__db);
+                    let __ingredients = <#jar_ty as salsa::storage::HasIngredientsFor< #ident >>::ingredient(__jar);
+                    let __id = __ingredients.#input_index.new_singleton_input(__runtime);
+
+                    #(
+                        __ingredients.#field_indices.store_new(__runtime, __id, self.#field_names.clone().ok_or_else(||"field not provided")?, durability);
```

Comment:
Thought about having `with_durabilities` too, but I dropped the idea because it sorts of requires the user to keep the order of fields in mind. I don't know if that cognitive overhead is okay. @MihailMihov, @nikomatsakis , @XFFXFF 
Though personally, I think it's better.

Another option I'm considering is having `with_fields` take a tuple, i.e. its `value` and it's `durability`. That localizes concerns. But, is taking tuples neat?

Reply:
> Thought about having with_durabilities too, but I dropped the idea because it sorts of requires the user to keep the order of fields in mind. I don't know if that cognitive overhead is okay.

The cognitive overhead point is something that I didn't consider, but I did go over the idea with the tuples and decided to not suggest it, because it wouldn't look very good if we needed to add more options for the fields in the future. And then we also lose the ability to change just 1 or 2 things about a field, we'd need to set all 5 (again this only applies if we do consider having more than just a value and a durability).

Also if tuples were an option all the intermediate structs produced by the builder would have an incomplete state, and the user would need to keep track of all the fields being set.

But that's just my opinion I think we should hear what others have to say.